crates/red_knot_python_semantic/src/semantic_index/usedef.rs

```diff
@@ -0,0 +1,137 @@
+use crate::ast_node_ref::AstNodeRef;
+use crate::db::Db;
+use crate::semantic_index::ast_ids::ScopedUseId;
+use crate::semantic_index::definition::Definition;
+use crate::semantic_index::symbol::{FileScopeId, ScopedSymbolId};
+use ruff_db::files::File;
+use ruff_index::IndexVec;
+use ruff_python_ast as ast;
+use salsa;
+
+/// An independently type-inferable expression.
+///
+/// Includes constraint expressions (e.g. if tests) and the RHS of an unpacking assignment.
+#[salsa::tracked]
+pub(crate) struct Expression<'db> {
+    /// The file in which the expression occurs.
+    #[id]
+    pub(crate) file: File,
+
+    /// The scope in which the expression occurs.
+    #[id]
+    pub(crate) scope: FileScopeId,
+
+    /// The statement node containing the expression.
+    /// We hold the statement node to clarify which specific contexts these
+    /// independently-inferable expressions can occur in, and limit the variants of this enum.
+    #[no_eq]
+    #[return_ref]
+    pub(crate) ast: ExpressionKind,
+}
+
+impl<'db> Expression<'db> {
+    /// Return the actual expression.
+    pub(crate) fn node(&self, db: &'db dyn Db) -> &ast::Expr {
+        match self.ast(db) {
+            ExpressionKind::AssignedValue(node_ref) => &node_ref.node().value,
+            ExpressionKind::IfTest(node_ref) => &node_ref.node().test,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) enum ExpressionKind {
+    AssignedValue(AstNodeRef<ast::StmtAssign>),
+    IfTest(AstNodeRef<ast::StmtIf>),
+    // TODO if-expressions, loops, etc
+}
+
+/// All definitions that can reach a given use of a name.
+#[derive(Debug)]
+pub(crate) struct UseDefMap<'db> {
+    // TODO store constraints with definitions for type narrowing
+    // TODO smallvecs?
+    /// Definitions that can reach a [`ScopedUseId`].
+    definitions_by_use: IndexVec<ScopedUseId, Vec<Definition<'db>>>,
+
+    /// Definitions of a symbol visible to other scopes.
+    public_definitions: IndexVec<ScopedSymbolId, Vec<Definition<'db>>>,
+}
+
+impl<'db> UseDefMap<'db> {
+    fn new() -> Self {
+        Self {
+            definitions_by_use: IndexVec::new(),
+            public_definitions: IndexVec::new(),
+        }
+    }
+
+    pub(crate) fn use_definitions(&self, use_id: ScopedUseId) -> &[Definition<'db>] {
+        &self.definitions_by_use[use_id]
+    }
+
+    pub(crate) fn public_definitions(&self, symbol: ScopedSymbolId) -> &[Definition<'db>] {
+        &self.public_definitions[symbol]
+    }
+}
+
+impl PartialEq for UseDefMap<'_> {
+    fn eq(&self, other: &Self) -> bool {
+        self.definitions_by_use == other.definitions_by_use
+            && self.public_definitions == other.public_definitions
+    }
+}
+
+impl Eq for UseDefMap<'_> {}
```

Comment:
Can we use `derive(PartialEq, Eq)`?

Reply:
Oh yeah, I took this from the symbol table, but there we manually implement PartialEq because we want to exclude a field from it. Will fix.